File: week 1 - C/problemSet1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Get the credit card number from the user
    long long card_number;
    do
    {
        card_number = get_long("Number: ");
    }
    while (card_number <= 0);

    // Calculate checksum using Luhn's algorithm
    int sum = 0;
    int digit_count = 0;
    long long temp = card_number;

    while (temp > 0)
    {
        int last_digit = temp % 10;

        if (digit_count % 2 == 1)
        {
            int doubled_digit = last_digit * 2;
            sum += (doubled_digit % 10) + (doubled_digit / 10);
        }
        else
        {
            sum += last_digit;
        }

        temp /= 10;
        digit_count++;
    }

    // Check the card's validity and its type
    if (sum % 10 == 0)
    {
        int first_digit = card_number / 10 % 10;
        int second_digit = card_number / 100 % 10;

        if ((digit_count == 15) && (second_digit == 3) && ((first_digit == 4) || (first_digit == 7)))
        {
            printf("AMEX\n");
        }
        else if ((digit_count == 16) && (second_digit >= 1) && (second_digit <= 5))
        {
            printf("MASTERCARD\n");
        }
        else if ((digit_count == 13 || digit_count == 16) && (first_digit == 4))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

File: Week 6 - Python/problemSet6/sentimentalCredit/credit.py
def main():
    # Get the credit card number from the user
    card_number = input("Number: ")

    if not card_number.isdigit() or int(card_number) <= 0:
        print("INVALID")
        return

    card_length = len(card_number)

    # Check the card's validity using Luhn's algorithm
    total = 0
    for i in range(card_length - 2, -1, -2):
        digit = int(card_number[i]) * 2
        if digit > 9:
            digit -= 9
        total += digit

    for i in range(card_length - 1, -1, -2):
        total += int(card_number[i])

    # Check the card type
    if total % 10 == 0:
        if card_length == 15 and card_number[:2] in ("34", "37"):
            print("AMEX")
        elif card_length == 16 and card_number[:2] in ("51", "52", "53", "54", "55"):
            print("MASTERCARD")
        elif (card_length == 13 or card_length == 16) and card_number[0] == "4":
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
